File: ABC/174/F.java
// editorial(TLE)
import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;

class BIT{
    int n;
    int a[];
    BIT(int n){
        this.n = n;
        a = new int[n+1];
    }
    void add(int i, int x){
        while(i <= n){
            a[i] += x;
            i += i&-i;
        }
    }
    int sum(int i){
        int ret = 0;
        while(i > 0){
            ret += a[i];
            i -= i&-i;
        }
        return ret;
    }
}

class Pair implements Comparable<Pair>{
    int l, r, ind;
    Pair(int l, int r, int i){
        this.l = l;
        this.r = r;
        this.ind = i;
    }

    @Override
    public int compareTo(Pair p){
        return this.r-p.r;
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String line[] = br.readLine().split(" ");
        int n = Integer.parseInt(line[0]);
        int q = Integer.parseInt(line[1]);
        line = br.readLine().split(" ");
        int a[] = new int[n];
        for(int i = 0; i < n; i++)  a[i] = Integer.parseInt(line[i])-1;
        ArrayList<Pair> arr = new ArrayList<Pair>(q);
        for(int i = 0; i < q; i++){
            line = br.readLine().split(" ");
            arr.add(new Pair(Integer.parseInt(line[0])-1, Integer.parseInt(line[1])-1, i));
        }
        arr.sort(null);
        int ans[] = new int[q];
        BIT bit = new BIT(n);
        int last[] = new int[n];
        Arrays.fill(last, -1);
        int qind = 0;
        for(int i = 0; i < n; i++){
            if(last[a[i]] != -1){
                bit.add(last[a[i]]+1, -1);
            }
            last[a[i]] = i;
            bit.add(last[a[i]]+1, 1);
            while(qind < q && arr.get(qind).r==i){
                ans[arr.get(qind).ind] = bit.sum(i+1)-bit.sum(arr.get(qind).l);
                qind++;
            }
        }

        for(int x : ans)    System.out.println(x);
    }
}

File: ABC/174/F.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

typedef long long ll;

struct BIT{
    // 1-origin
    vector<int> v;
    int n;

    BIT(int _n):
        v(vector<int>(_n+1, 0)), n(_n) {}

    int sum(int i){
        int s = 0;
        while(i > 0){
            s += v[i];
            i -= lsb(i);
        }
        return s;
    }
    
    int sum(int l, int r){
        return sum(r) - sum(l-1);
    }

    void add(int i, int x){
        while(i <= n){
            v[i] += x;
            i += lsb(i);
        }
    }

    ll invNum(){
        ll ret = 0;
        for(int j = 0; j < n; j++){
            ret += j - sum(v[j]);
            add(v[j], 1);
        }
        return ret;
    }

private:
    // least significant bit
    int lsb(int i){
        return i & -i;
    }
};

typedef pair<int,int> pii;

int main(){
    cin.tie(0);
    ios::sync_with_stdio(false);
    int n, q;
    cin >> n >> q;
    int c[n];
    for(int i = 0; i < n; i++)  cin >> c[i], c[i]--;
    vector<pair<pii,int>> vp;
    for(int i = 0; i < q; i++){
        int l, r;
        cin >> l >> r;
        vp.push_back({{--r, --l}, i});
    }
    sort(vp.begin(), vp.end());
    vector<int> last(n, -1), ans(q);
    int qind = 0;
    BIT bit(n);
    for(int i = 0; i < n; i++){
        if(last[c[i]] != -1){
            bit.add(last[c[i]]+1, -1);
        }
        last[c[i]] = i;
        bit.add(last[c[i]]+1, 1);
        while(qind < q && vp[qind].first.first == i){
            ans[vp[qind].second] = bit.sum(i+1)-bit.sum(vp[qind].first.second);
            qind++;
        }
    }
    for(int i = 0; i < q; i++)  cout << ans[i] << endl;
    return 0;
}

